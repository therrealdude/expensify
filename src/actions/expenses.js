import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense) => {
  return ({
  type: 'ADD_EXPENSE',
  expense
})};

export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
      expenses.push({
        id: childSnapshot.key,
        ...childSnapshot.val()
      });
     });
     dispatch(setExpenses(expenses));
   });
 }
};

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description,
      note,
      amount,
      createdAt
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const startRemoveExpense = (id) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then((ref) => {
      dispatch(removeExpense({ id }));
    });
  };
};

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then((ref) => {
      dispatch(editExpense(id, updates));
    })
  }
}
