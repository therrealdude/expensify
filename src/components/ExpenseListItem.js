import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <li>
    <div>
      <p>Description: <Link to={`/edit/${id}`}>{description}</Link></p>
      <p>Amount: {numeral(amount / 100).format('$0,000.00')}</p>
      <p>Created At: {moment(createdAt).format('MMMM Do, YYYY')}</p>
    </div>
  </li>
);

export default ExpenseListItem;
