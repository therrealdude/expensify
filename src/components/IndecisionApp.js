import React from 'react';
import AddOption from './AddOption.js'
import Header from './Header.js';
import Action from './Action.js';
import Options from './Options.js';
import RemoveAll from './RemoveAll.js';
import OptionModal from './OptionModal.js';

class IndecisionApp extends React.Component {
  state = {
    title: 'Indecision',
    subtitle: 'Put your life in the hands of a computer.',
    options: [],
    selectedOption: false
  };
  handlePick = () => {
    const randomNum = Math.floor(Math.random() * this.state.options.length);
    const option = this.state.options[randomNum];
    this.setState(() => ({ selectedOption: this.state.options[randomNum] }))
  };
  handleDeleteOptions = () => {
    this.setState(() => ({ options: [] }));
  };
  handleDeleteOption = (option) => {
    this.setState((prevState) => ({
      options: prevState.options.filter((o) => o !== option)
    }))
  };
  handleAddOption = (option) => {
      console.log(option);
      if (!option){
        return 'Enter valid value to add item';
      } else if (this.state.options.indexOf(option) > -1) {
        return 'This option already exists';
      }
      console.log(this.state.options);
      this.setState((prevState) => ({
          options: prevState.options.concat(option)
      }));
  };
  handleCloseModal = () => {
    this.setState(() => ({ selectedOption: undefined }))
  };
  componentDidMount(){
    try{
      const json = localStorage.getItem('options');
      const options = JSON.parse(json);
      if (options){
        this.setState(() => ({ options }));
      }
    }
    catch (e){
      //Do nothing
    }
  }
  componentDidUpdate(prevProps, prevState){
    if (prevState.options.length !== this.state.options.length) {
      const json = JSON.stringify(this.state.options);
      localStorage.setItem('options', json);
    }
  }
  componentWillUnmount(){
    console.log('componentWillUnmount');
  }

  render() {
    return (
      <div>
        <Header subtitle={this.state.subtitle} />
        <div className="container">
          <Action hasOptions={this.state.options.length > 0} handlePick={this.handlePick}/>
          <Options options={this.state.options}
            handleDeleteOption={this.handleDeleteOption}/>
          <AddOption  handleAddOption={this.handleAddOption}/>
          <RemoveAll options={this.state.options} handleDeleteOptions={this.handleDeleteOptions}/>
          <OptionModal selectedOption={this.state.selectedOption} handleCloseModal={this.handleCloseModal}/>
        </div>
      </div>
    );
  }
}

IndecisionApp.defaultProps = {
  options: []
};

export default IndecisionApp;
