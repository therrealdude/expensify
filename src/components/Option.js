import React from 'react';

export default Option = (props) => (
    <div>
      <p>{props.option}</p>
      <button className="button button--link" onClick={(e) => {
        props.handleDeleteOption(props.option)
      }}>remove</button>
    </div>
  );
