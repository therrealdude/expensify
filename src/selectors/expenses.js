import moment from 'moment';

export const selectExpenses = (expenses, { text, sortBy, startDate, endDate }) =>  expenses.filter((expense) => {
  const createdAtMoment = moment(expense.createdAt);
  const startDateMatch = !!startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
  const endDateMatch = !!endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
  const textMatch = text === undefined || !expense.description || expense.description.toLowerCase().includes(text.toLowerCase());

  return startDateMatch && endDateMatch && textMatch;
}).sort((a, b) => {
  if (sortBy === 'date')
  {
    return a.createdAt < b.createdAt ? 1 : -1;
  }
  else if (sortBy === 'amount'){
    return a.amount < b.amount ? 1 : -1;
  }
});

export const totalExpenses = (expenses) =>
  expenses.map(expense => expense.amount).reduce(
    (a, b) =>
      a + b,
      0);
