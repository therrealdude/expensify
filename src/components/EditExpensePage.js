import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (id, expense) => {
    this.props.editExpense(id, expense);
    this.props.history.push('/');
  };
  onRemoveExpense = (id) => {
    this.props.removeExpense(id);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <ExpenseForm
          onSubmit={this.onSubmit}
          expense={this.props.expense} />
        <button onClick={this.onRemoveExpense}>Remove</button>
      </div>
    );
  }
}

const mapDispathToProps = (dispatch) => ({
    editExpense: (id, expense) =>  dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense({ id }))
  });

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
};

export default connect(mapStateToProps, mapDispathToProps)(EditExpensePage);
