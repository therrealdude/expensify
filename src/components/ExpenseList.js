import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import { selectExpenses } from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    { props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      <ul>
        {props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)}
      </ul>
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
}

export default connect(mapStateToProps)(ExpenseList);
