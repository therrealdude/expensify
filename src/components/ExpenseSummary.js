import React from 'react';
import { totalExpenses, selectExpenses } from '../selectors/expenses';
import numeral from 'numeral';
import { connect } from 'react-redux';

export const ExpenseSummary = (props) => (
  <div>
    <p>Viewing {props.expenses.length} expense{ props.expenses.length === 1 ? '' : 's'} totaling {numeral(totalExpenses(props.expenses) / 100).format('$0,000.00')}</p>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
}

export default connect(mapStateToProps)(ExpenseSummary);
