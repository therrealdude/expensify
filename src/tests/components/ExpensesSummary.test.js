import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test('should render ExpenseSummary with expenses', () => {
  const wrapper = shallow(<ExpenseSummary expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});
