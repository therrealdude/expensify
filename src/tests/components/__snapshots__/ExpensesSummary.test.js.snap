// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with expenses 1`] = `
<div>
  <p>
    Viewing 
    3
     expense
    s
     totaling 
    $1,141.95
  </p>
</div>
`;
